### PR TITLE
Use version.hpp for package versions 

### PR DESCRIPTION
**Pull request by danieljurek**
Uses version.hpp: 

* Tagging repository ([link](https://github.com/Azure/azure-sdk-for-cpp/releases/tag/azure-template_1.0.0-beta.1))
* Publishing docs ([link](https://azuresdkdocs.blob.core.windows.net/$web/cpp/azure-template/1.0.0-beta.1/index.html))
* Incrementing and opening PR ([link](https://github.com/Azure/azure-sdk-for-cpp/pull/571))

Core and Template will be able to publish. 

There is an exception for storage here. To unblock storage release I added a condition in `Get-PkgVersion.ps1` that will fall back to a version.txt file. When storage is using version.hpp like core and template they will automatically get the same benefits that core and template have in this PR. Today storage releases will be able to successfully release docs and create the release tag on GitHub but the version incrementing script will not work. 

Closing https://github.com/Azure/azure-sdk-for-cpp/pull/567 because this kind of engsys work is best done from a branch in the main repo instead of a PR.

### CODE REVIEW

File: eng/scripts/Get-PkgVersion.ps1 [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    [string] $PackageName
+)
+
+. ${PSScriptRoot}/SdkVersion-Common.ps1
+
+$versionFileLocation = Get-VersionHppLocaiton `
+    -ServiceDirectory $ServiceDirectory `
+    -PackageName $PackageName
+
+if (!$versionFileLocation) {
```
Comment (danieljurek) [inline eng/scripts/Get-PkgVersion.ps1:16]: Falls back to unblock storage artifact generation (docs, changelog, tag metadata)

File: cmake-modules/az_version.cmake [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+# constructed from version.hpp. The version.hpp must follow the example in
+# templates for version information to parse correctly.
+
+function(get_az_version VERSION_HPP_FILE)
+    if(NOT EXISTS ${VERSION_HPP_FILE})
+        message(FATAL_ERROR "Missing Version file ${VERSION_HPP_FILE}")
+    endif()
+
+    file (STRINGS ${VERSION_HPP_FILE} VERSION_H_CONTENT NEWLINE_CONSUME)
+    message(STATUS "Read version from ${VERSION_HPP_FILE}")
```
Comment (RickWinter) [inline cmake-modules/az_version.cmake:11]:
```suggestion
    message(STATUS "Retrieving version from ${VERSION_HPP_FILE}")
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 3]
```diff
@@ -132,8 +132,7 @@ jobs:
           - pwsh: |
               $outputPath = Join-Path -Path $(Build.ArtifactStagingDirectory) packages/${{ artifact.Name }}/package-info.json
 
-              # TODO: Read version from a .h file instead of version.txt
-              $version = Get-Content -Path sdk/${{ parameters.ServiceDirectory }}/${{ artifact.Path }}/version.txt
+              $version = eng/scripts/Get-PkgVersion -ServiceDirectory ${{ parameters.ServiceDirectory }} -PackageName ${{ artifact.Path }}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:135]: Does PS allow you to call a script without an extension?

File: eng/scripts/SdkVersion-Common.ps1 [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# NOTE: Update-PkgVersion and Get-PkgVersion relies on these variables and
+# functions
+$RepoRoot = "${PSScriptRoot}/../.."
+$VersionRegex = '(#define AZURE_\w+_VERSION_MAJOR )(?<major>[0-9]+)(\s+#define AZURE_\w+_VERSION_MINOR )(?<minor>[0-9]+)(\s+#define AZURE_\w+_VERSION_PATCH )(?<patch>[0-9]+)(\s+#define AZURE_\w+_VERSION_PRERELEASE )"(?<prerelease>[a-zA-Z0-9.]*)"';
```
Comment (weshaggard) [inline eng/scripts/SdkVersion-Common.ps1:4]: We should consider using the regex defined in the SemVer.ps1 script. We can add the extra parts around it if needed.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 3 above)
Comment (danieljurek) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:135]: This is a typo, however it does work.

File: eng/scripts/SdkVersion-Common.ps1 (hunk 4 above)
Comment (danieljurek) [reply eng/scripts/SdkVersion-Common.ps1:4]: https://github.com/Azure/azure-sdk-for-cpp/issues/609

**Pull request merged**